### PR TITLE
feat(i18n): able to finetune transliterator

### PR DESCRIPTION
**Pull request by BLumia**
**Changes**
After https://github.com/jellyfin/jellyfin/pull/11172, this gives user the ability to finetune the behavior of the existing transliterator to fit their own need. All they need to do is set the `JELLYFIN_TRANSLITERATOR_ID` environment variable to a preferred value, which is explained in [this page](https://unicode-org.github.io/icu/userguide/transforms/general/rules.html#script-transliterators) and [this page](https://unicode-org.github.io/icu/userguide/transforms/general/).

This approach is more flexible than only offers an option to disable the ICU transliterator, users can then:

1. Only transliterate specified curtain languages instead of "Any", for example, use `Hangul-Latin; Hiragana-Latin;` instead of `Any-Latin;` so only Hangul and Hiragana characters will get transliterated. Other non-ascii characters (e.g. Kanji and Chinese characters) will be untouched.
2. Finetune transliterator behavior (since you can set any valid Transliterator ID to that environment variable).
3. Simply disable it. Set this environment variable with an empty value, the transliterator will simply do nothing.

Some side notes (some of the notes have nothing to do with the changes made in this PR though):

1. Users (and plugins as well) can always override the sorting title by setting the sort title field manually. If a sort title is already manually set, Jellyfin won't attempt to override it.
2. If you choose to set sort title manually or by plugin, make sure at least the first character of the sort title is an ascii character, or you'll not able to find that item by using the alphabet filter on the right-hand side of the page.
3. When generating the fallback sort title, Jellyfin already attempt to remove diacritics in `ModifySortChunks` Before the ICU transliterator gets used. Keep that in mind when choosing the right transliterator rule (Transliterator ID).
4. You need to trigger a re-scan to make Jellyfin update the existing sort title cache. Choosing "Update Missing Metadata" will be fine.

**Issues**

- Fixes https://github.com/jellyfin/jellyfin/issues/11880
- See also: https://github.com/jellyfin/jellyfin/pull/11172#issuecomment-2249687663

### CODE REVIEW

**Comment by gnattu** [comment]
Related: https://github.com/jellyfin/jellyfin/issues/11880

File: src/Jellyfin.Extensions/StringExtensions.cs [hunk 1]
```diff
@@ -9,8 +9,21 @@ namespace Jellyfin.Extensions
     /// </summary>
     public static partial class StringExtensions
     {
-        private static readonly Lazy<Transliterator> _transliterator = new(() => Transliterator.GetInstance(
-            "Any-Latin; Latin-Ascii; Lower; NFD; [:Nonspacing Mark:] Remove; [:Punctuation:] Remove;"));
+        private static readonly Lazy<string> _transliteratorId = new(() =>
+            Environment.GetEnvironmentVariable("JELLYFIN_TRANSLITERATOR_ID")
```
Comment (gnattu) [inline src/Jellyfin.Extensions/StringExtensions.cs:13]: I'm personally fine with reading env var directly but I'm not sure if there is any better way to integrate the configuration for string extensions

**Review by gnattu — APPROVED** [review]
Need better documentation, ideally on jellyfin.org. Probably also should mention this is in changelog as user actions needed to fix undesired behaviors. Otherwise LGTM

**Comment by BLumia** [comment]
> Need better documentation, ideally on jellyfin.org.

Agree.

Related to this, maybe we should simply change `Sort title` to something else (e.g. "Transliterated title"), too? By doing this, user could understand the `sorttitle` metadata is supposed to be ascii-only, and maybe more plugin developer can make use of that metadata.

**Comment by gnattu** [comment]
> sortname metadata is supposed to be ascii-only

Actually it is not. Computers do sort non ascii characters but may or may not in the human acceptable way and using ascii only characters is a easier way to control sorting behavior. But if users really want to input a non-ascii character it is also OK.

**Comment by BLumia** [comment]
> Computers do sort non ascii characters but may or may not in the human acceptable way

Is there any real-life usage example to help me understand better about this usage? If it's intended to allow `sortname` contain non-ascii character, then why user cannot use the filter on the right-hand side to find there programs which their `sortname`s are started with non-ascii characters?

**Comment by gnattu** [comment]
> > Computers do sort non ascii characters but may or may not in the human acceptable way
> 
> Is there any real-life usage example to help me understand better about this usage? If it's intended to allow `sortname` contain non-ascii character, then why user cannot use the filter on the right-hand side to find there programs which their `sortname`s are started with non-ascii characters?

> If it's intended to allow `sortname` contain non-ascii character, then why user cannot use the filter on the right-hand side to find there programs which their `sortname`s are started with non-ascii characters?

Because there are too many of them

> Is there any real-life usage example to help me understand better about this usage?

For example the ascii order of romaji of hiragana and katakanas are not in the same order as the Japanese traditional あいうえお order. Language is just complicated and you cannot assume the ASCII order is the most correct order.

**Comment by BLumia** [comment]
> Because there are too many of them

If a user has too many items which starts with a non-ascii character, then I'd say the right-hand side filter is completely useless for such user. Things like [Alphabetic Index](https://icu.unicode.org/design/alphabetic-index) *might* be helpful to improve this, while I'm not 100% sure about it. Whatever if that can be improved, it will be a completely new discussion topic which is not related to this PR.

Anyway, related to this PR, as a conclusion, we won't change the copywriting of "Sort Title" to something else :)

**Review by crobibero — APPROVED** [review]

**Comment by harry71019** [comment]
I tried setting `JELLYFIN_TRANSLITERATOR_ID` to nothing but that doesn't seem to change/disable the original settings. So I set it to `Hangul-Latin` since I don't have any hangul.

**Comment by gnattu** [comment]
> I tried setting `JELLYFIN_TRANSLITERATOR_ID` to nothing but that doesn't seem to change/disable the original settings. So I set it to `Hangul-Latin` since I don't have any hangul.

It needs to be explicitly empty string, not really "nothing". Really nothing is the same as use default.

**Comment by harry71019** [comment]
> > I tried setting `JELLYFIN_TRANSLITERATOR_ID` to nothing but that doesn't seem to change/disable the original settings. So I set it to `Hangul-Latin` since I don't have any hangul.
> 
> It needs to be explicitly empty string, not really "nothing". Really nothing is the same as use default.

Thanks for the explanation!

**Comment by trustin** [comment]
`JELLYFIN_TRANSLITERATOR_ID=Any-Hex` transliterates everything into hexadecimal representation of unicode value, which largely fixes my issue.

**Pull request merged**